docs/architecture/adr-071-cryptography-v2.md

```diff
@@ -0,0 +1,569 @@
+
+    Wallet->>Wallet: Init(Keyring)
+
+
+    Wallet->>Application: Wallet Instance
+```
+
+***Signing and verifying a message***
+
+```mermaid
+sequenceDiagram
+    participant Application
+    participant Wallet
+    participant Keyring
+    participant CryptoProvider
+    participant Signer
+    participant Verifier
+
+    Application->>Wallet: GetSigner(Address)
+    Wallet->>Keyring: GetCryptoProvider(ItemId)
+    Keyring->>SecureStorage: Get(ItemId.Uuid)
+    SecureStorage->>Keyring: SecureItem
+    Keyring->>Keyring: GetBuilderFunction(ItemId.Uuid)
+    Keyring->>CryptoProvider: Build(SecureItem)
+    CryptoProvider->>Wallet: CryptoProvider instance
+    Wallet->>CryptoProvider: GetSigner()
+    CryptoProvider->>Application: Signer instance
+    Application->>Signer: Sign()
+    Signer->>Application: Signed message
+    Application->>Wallet: GetVerifier(address)
+    Wallet->>CryptoProvider: GetVerifier()
+    CryptoProvider->>Wallet: Verifier instance
+    Application->>Verifier: Verify()
+    Verifier->>Application: true/false
+```
+
+## Alternatives
+
+The alternatives may vary in the way of distributing the packages, grouping them together as for example verify and signing in
+one place. This will affect the granularity of the code, thus the reusability and modularity. We aim to balance between simplicity and
+granularity.
+
+## Decision
+
+We will:
+
+* Refactor module structure as described above.
+* Define types and interfaces as the code attached.
+* Refactor existing code into new structure and interfaces.
+* Implement Unit Tests to ensure no backward compatibility issues.
+
+## Consequences 
+
+### Impact on the SDK codebase
+
+This ADR primarily impacts the client-side components of the Cosmos SDK. The consensus layer of the Cosmos SDK, which Tendermint Core largely handles, will not be directly affected by these changes. The cryptographic tools, such as hashers and verifiers, will still be available as external packages. These can be imported and utilized directly in the consensus code without the need for a CryptoProvider instance. This approach ensures that the consensus layer remains decoupled from the specific implementations of cryptographic operations. It maintains the integrity and stability of the consensus process while allowing for enhancements and extensions in the cryptographic functionalities.
+
+### Backwards Compatibility
+
+Some packages will need a medium to heavy refactor to be compatible with this ADR. 
+In short, packages using _Keyring_ (current SDK) will need to be adapted to use the new Keyring and CryptoProvider interfaces.
+Other special cases where a refactor will be needed, are the ones that make use crypto components in isolation like the  _PrivateKey_ and _PublicKey_ structs
+to sign and verify transactions respectively.
+
+As first approach, the most affected packages are:
+- crypto/types
+- client/rpc
+- client/tx
+- client/keys
+- types/tx/signing
+- x/auth
+- x/auth/client
+- x/slashing
```

Comment:
if these packages will be affected then we are missing a design in how will these modules get the needed info from the new cryptoprovider. will it get passed down to modules?

Reply:
yep, we have that covered (not in detail for every package though) [here](https://github.com/cosmos/cosmos-sdk/blob/fb0fba596a34dab665474a39022dba53ee2eb6af/docs/architecture/adr-071-cryptography-v2.md?plain=1#L473-L476).
But the idea is if:
- package uses `keyring`: direct replacement with the new keyring that uses CryptoProvider under the hood
- package signs without using `keyring`: this one will be the most affected. To avoid bleeding privkeys, these pkgs should be refactored to use keyring instead.
- package uses any other crypto tool than sign (hash, enc/dec, etc): [not affected](https://github.com/cosmos/cosmos-sdk/blob/fb0fba596a34dab665474a39022dba53ee2eb6af/docs/architecture/adr-071-cryptography-v2.md?plain=1#L469).

Reply:
tx and signature verification is affected not as a whole but only on the cryptography related stuff. Our proposal aims to decouple stuff. May I ask, how do you see it merging client and state machine?

Reply:
@tac0turtle friendly reminder over this discussion

Reply:
@tac0turtle cc

Reply:
i dont see how things like slashing are affected by this ADR. From my understanding this adr is client only, but the adr mentioned slashing and other modules being affected. This is contradicting and leaving me more confused on the goals and design of this pr

Reply:
The `x/slashing` module is not directly affected by this ADR, but it might be indirectly impacted. This is because several .go files in that package import [crypto/types](https://github.com/cosmos/cosmos-sdk/blob/86eaeed5199efb2871f49d1928ddbc4dff19edbb/x/slashing/keeper/keeper.go#L15). If these types change, even just their names, it will require changes in that package. The same applies to other packages, mostly in the keeper.go files.

However, there's always the option to maintain the v1 pub/priv-key style for backward compatibility. But considering all the [advantages](https://github.com/cosmos/cosmos-sdk/pull/18824#discussion_r1440791183) of the new v2 `BaseKey` interface, it might be a good idea to migrate every usage to v2 in subsequent PRs.

This is why we considered appropriate to include these packages in that list, not as an inevitable but a possible impact.